### ISSUE TITLE
Trying to compile node_pcap with npm script

### ISSUE THREAD

**Issue by fridaystreet** [nneves/openwrt_mips_ar9331_nodejs#3]
Hi,

I'm trying to use the npm script to cross compile node_pcap for MIPS, but I just get the output below. Not really sure what's going on, but the directory /home/system/openwrt_mips_ar9331_nodejs/node_modules/pcap/ doesn't seem to exist after running the script. 

It appears that it is being deleted at some point in the process. I have tried copying the pcap directory into mode_modules dir and rerunning the script, but I get the same error, then after the directory no longer exists.

Any assistance you might be able to provide would be greatly appreciated

Cheers
Paul

Cross-compile NPM package for OpenWrt MIPS AR9331 SoC: npm install pcap
npm http GET https://registry.npmjs.org/pcap
npm http 304 https://registry.npmjs.org/pcap
npm http GET https://registry.npmjs.org/nan
npm http 304 https://registry.npmjs.org/nan

> socketwatcher@0.2.1 install /home/system/openwrt_mips_ar9331_nodejs/node_modules/pcap/node_modules/socketwatcher
> node-gyp rebuild

make: Entering directory `/home/system/openwrt_mips_ar9331_nodejs/node_modules/pcap/node_modules/socketwatcher/build'
  CXX(target) Release/obj.target/socketwatcher/socket_watcher.o
  SOLINK_MODULE(target) Release/obj.target/socketwatcher.node
  COPY Release/socketwatcher.node
make: Leaving directory`/home/system/openwrt_mips_ar9331_nodejs/node_modules/pcap/node_modules/socketwatcher/build'

> pcap@2.0.1 install /home/system/openwrt_mips_ar9331_nodejs/node_modules/pcap
> node-gyp rebuild

make: Entering directory `/home/system/openwrt_mips_ar9331_nodejs/node_modules/pcap/build'
  CXX(target) Release/obj.target/pcap_binding/pcap_binding.o
../pcap_binding.cc:2:23: fatal error: pcap/pcap.h: No such file or directory
compilation terminated.
make: *** [Release/obj.target/pcap_binding/pcap_binding.o] Error 1
make: Leaving directory`/home/system/openwrt_mips_ar9331_nodejs/node_modules/pcap/build'
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:267:23)
gyp ERR! stack     at ChildProcess.EventEmitter.emit (events.js:98:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:789:12)
gyp ERR! System Linux 3.2.0-29-generic-pae
gyp ERR! command "node" "/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /home/system/openwrt_mips_ar9331_nodejs/node_modules/pcap
gyp ERR! node -v v0.10.17
gyp ERR! node-gyp -v v0.10.9
gyp ERR! not ok
npm ERR! weird error 1
npm ERR! not ok code 0

**Comment by fridaystreet** [comment]
I got a little bit further by realising in some of the .cc file in pcap module had includes to pcap/pcap.h which if I changes to the openwrt source path was able to get past those errors and build the files. This just left one last error

/usr/bin/ld: cannot find -lpcap

I googled this and someone mentioned for a different module they had removed the reference in config.gypi. It wasn't in that file it was in binding.gyp. I removed it from the libraries section of that and the build seem to complete ok. 

My problem now is that I can't seem to get it to run on my mips device. I've copied the node_modules directory from my openwrt build machine to my project directory on the mips device, but now I'm just getting the error below. 

I really need to get this working and will happily pay you if you are able to assist me?

Cheers
Paul

module.js:340
    throw err;
          ^
Error: Cannot find module './build/Release/pcap_binding'
    at Function.Module._resolveFilename (module.js:338:15)
    at Function.Module._load (module.js:280:25)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at Object.<anonymous> (/tmp/freedom/node_modules/pcap/pcap.js:3:21)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)

**Comment by nneves** [comment]
Hi Paul,

let me make some tests first, not sure if the libpcap required dependency is from the build-system OS or from the OpenWrt itself, I'm more leaned to the last, if so I'm not sure how to link it in the cross-compile procedure! I've previously cross-compiled a C binary for openwrt that required some libs to be linked, but this is something that I haven't tried in this process!

**Comment by fridaystreet** [comment]
Hi Nelson,

Thanks for getting back to me. I realise it isn't really your script at fault, but if you are able to assist it would be greatly appreciated. Like I said I'm happy to pay for your time if necessary.

Cheers
Paul

**Comment by fridaystreet** [comment]
Got a bit further still. I don't think it ran properly the first time I ran it without the -lpcap removed. Even though node-gyp rebuild appeared to complete ok, it was still missing the pcap_binding.node file in the Release folder.

I've run it again and that file was created ok this time and running it on my device I know get the following.

module.js:356
  Module._extensions{extension}(this, filename);"     

 (note the braces should be square brackets, but the markup was picking them up so I changed them)

Error: File not found
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at Object.<anonymous> (/tmp/freedom/node_modules/pcap/pcap.js:3:21)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)

Which looks as though the module at least is loading. A google search of this error seems to indicate incompatibility with the version of node. I'm wondering if v0.10.5 is too old?

Cheers
Paul

**Comment by fridaystreet** [comment]
Managed to resolve the issue as far as getting the pcap module to run. It was missing libstdc++.so.6.0.16. So completely unrelated to this apologies.

My only issue now is more related to node_pcap itself or more so libpcap on openwrt. I think I need libpcap-dev as I'm getting various different can't resolve symbol errors for different pcap functions, which I'm guesing are in libpcap-dev and not in the standard libpcap.

Oh well thanks for your help anyway. I'll have to keep digging to see if I can find or build a libpcap-dev for openwrt on mips.

Cheers
Paul

**Comment by nneves** [comment]
Hi Paul,

nice to see you made some progress, I had some other problems and got stuck in an early stage!
I have some tips for your current status, let me know if this helps! I think that you will only require to cross-compile the libpcap from OpenWrt, here are the steps to follow:

```
cd openwrtsource
make menuconfig

-> navigate to:
Libraries

-> select library (double space for built-in library: <*>)
<*> libpcap Low-level packet capture library

make V=99
find ./ | grep “libpcap”
```

you can track down the compiled libs here:
openwrtsource/staging_dir/target-mipsel_r2_uClibc-0.9.33.2/usr/lib

libpcap.so.1.1.1 is around 216Kb and should be copied to your OpenWrt system (as you did for the libstdc++.so.6.0.16).

Also, there is a libpcap.a (318Kb), this is only required when you are cross-compiling stuff that require static linking (remember the -lpcap flag you removed, it would probably required this lib somewhere, since there is no static linking it will require to dynamically link with the .so version during runtime)

So, follow the above steps to cross-compile libpcap for OpenWrt, transfer the .so lib to your embedded system, if it still can't find the lib, you may need to create a symbolic link for the one it expects to (instead of *.so.1.1.1 to *.so)

Not sure if this will help, you are far beyond something I have ever tried! But looks great ;)

**Comment by fridaystreet** [comment]
Hi Nelson,

Thanks for the info. My knowledge of c and compiling is pretty limited. I have managed to get it working. I needed to copy the libpcap library from the packages directory into the staging directory below then ld -lpcap was able to find the pcap library and it built ok.

/openwrtsource/staging_dir/toolchain-mips_r2_gcc-4.6-linaro_uClibc-0.9.33.2/lib/

My issue now is that when I try and run any js scripts to use pcap I get the following error which apparently is an out of memory issue. 

terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
Aborted

I can run node ok and I've tried running some basic scripts to simulate load and they work fine. The device only has 64mb RAM, but I'm hoping it's something else. I've raised an issue on the node_pcap page to see if they may be able to help https://github.com/mranney/node_pcap/issues/165

Cheers
Paul
